scrapy/linkextractors/__init__.py

```diff
@@ -101,7 +101,7 @@ def _process_links(self, links):
         links = [x for x in links if self._link_allowed(x)]
         if self.canonicalize:
             for link in links:
-                link.url = canonicalize_url(urlparse(link.url))
+                link.url = canonicalize_url(link.url)
```

Comment:
I'm also not 100% sure, but for me it looks like SgmlLinkExtractor did that as a performance optimization, in order to avoid parsing URL multiple times. But now there is no performance benefit, and documented canonicalize_url API is to accept URLs as strings.